### PR TITLE
fix(api-rest): refactor ajax method to not relying on side effects

### LINKED ISSUE #11480: Graphql/Appsync api returns "BadRequestException"/"Signature expired" since aws-amplify 5.2.1 if client's clock i 5 Minutes off
### Before opening, please confirm:

- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-js/issues?q=is%3Aissue+) and [discussions](https://github.com/aws-amplify/amplify-js/discussions).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### JavaScript Framework

Angular

### Amplify APIs

GraphQL API

### Amplify Categories

api

### Environment information

<details>

```
# Put output below this line
System:
    OS: Linux 6.2 Ubuntu 23.04 23.04 (Lunar Lobster)
    CPU: (12) x64 12th Gen Intel(R) Core(TM) i7-1265U
    Memory: 31.48 GB / 46.71 GB
    Container: Yes
    Shell: 5.9 - /usr/bin/zsh
  Binaries:
    Node: 18.13.0 - ~/.nvm/versions/node/v18.13.0/bin/node
    Yarn: 3.2.2 - ~/.nvm/versions/node/v18.13.0/bin/yarn
    npm: 8.19.3 - ~/.nvm/versions/node/v18.13.0/bin/npm
  Browsers:
    Chrome: 114.0.5735.106
  npmPackages:
    @tsconfig/node18: ^1.0.3 => 1.0.3 
    @types/node: ^18.16.16 => 18.16.16 (18.16.0, 18.16.8)
    @typescript-eslint/eslint-plugin: ^5.59.9 => 5.59.9 
    @typescript-eslint/parser: ^5.59.9 => 5.59.9 (5.59.1)
    eslint: ^8.42.0 => 8.42.0 
    eslint-config-prettier: ^8.8.0 => 8.8.0 
    eslint-plugin-prettier: ^4.2.1 => 4.2.1 
    nx: 16.3.2 => 16.3.2 (16.2.2)
    nx-cloud: 16.0.5 => 16.0.5 
    prettier: ^2.8.8 => 2.8.8 
    prettier-plugin-jsdoc: 0.4.2 => 0.4.2 
    ts-node: ^10.9.1 => 10.9.1 
    tsconfig-paths: ^4.2.0 => 4.2.0 (3.14.2)
    typescript: 5.0.4 => 5.0.4 
  npmGlobalPackages:
    @githubnext/github-copilot-cli: 0.1.20
    corepack: 0.15.2
    npm: 8.19.3


```

</details>

### Describe the bug

When setting the clock in my OS more than five minutes in the past, all graphql requests return with:

> {
>   "errors" : [ {
>     "errorType" : "BadRequestException",
>     "message" : "Signature expired: 20230609T080915Z is now earlier than 20230609T082206Z (20230609T082706Z - 5 min.)"
>   } ]
> }

After trying multiple versions I found that this happens since amplify version 5.2.1.

As a result amplify also throws an error:

> TypeError: Cannot read properties of undefined (reading 'match')
>     at Object.getDateFromHeaderString (http://localhost:4200/vendor.js:11048:67)
>     at http://localhost:4200/vendor.js:988:94
>     at _ZoneDelegate.invoke (http://localhost:4200/polyfills.js:8256:160)
>     at Object.onInvoke (http://localhost:4200/vendor.js:172897:25)
>     at _ZoneDelegate.invoke (http://localhost:4200/polyfills.js:8256:48)
>     at Zone.run (http://localhost:4200/polyfills.js:8027:37)
>     at http://localhost:4200/polyfills.js:9174:28
>     at _ZoneDelegate.invokeTask (http://localhost:4200/polyfills.js:8283:173)
>     at http://localhost:4200/vendor.js:172584:49
>     at AsyncStackTaggingZoneSpec.onInvokeTask (http://localhost:4200/vendor.js:172584:30)

### Expected behavior

Amplify should handle a drifted clock as before 5.2.1.

### Reproduction steps

- Setup an Amplify app with auth and a appysnc api
- Set the clock more than 5 Minutes in the past
- Try to query something from the API and check the response and thrown errors 

### Code Snippet

```javascript
// Put your code below this line.

```

### Log output

<details>

```
// Put your logs below this line


```

</details>

### aws-exports.js

_No response_

### Manual configuration

_No response_

### Additional configuration

_No response_

### Mobile Device

_No response_

### Mobile Operating System

_No response_

### Mobile Browser

_No response_

### Mobile Browser Version

_No response_

### Additional information and screenshots

_No response_

### PR DESCRIPTION
**Pull request by AllanZhengYP**
This fix is for #11480

The previous Signer.sign() implementation not only returns signed request, but also sets the signed headers to input request. However this change breaks RestClient.ajax() clock skew correction which relies on the x-amz-date header set by signer to the input request object to indicate the current client side time.

With this fix, the `RestClient.ajax()` no longer rely on the side effect of signer. 

RestClient accesses the `x-amz-date` header that set to the input request by signer:
https://github.com/aws-amplify/amplify-js/blob/0fa40e6efb67def43f0eb912b3f2ed34da883f5f/packages/api-rest/src/RestClient.ts#L186-L206
The `requestDate` was later used to compare with the Date header from the server side to populate the clock skew.

#### Description of how you validated changes
Manually reproduced the issue and made sure the fix produced the same headers compared to previous implementation.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by cshfang — APPROVED** [review]

File: packages/api-rest/src/RestClient.ts [hunk 1]
```diff
@@ -180,39 +180,42 @@ export class RestClient {
...
+			logger.debug('No credentials available, the request will be unsigned');
+			return this._request(params, isAllResponse);
+		}
+		let signedParams;
+		try {
+			signedParams = this._sign({ ...params }, credentials, {
+				region,
+				service,
+			});
+			const response = await axios(signedParams);
```
Comment (elorzafe) [inline packages/api-rest/src/RestClient.ts:196]: Why we need to call `axios` here and not `this.ajax`?

Comment (AllanZhengYP) [reply packages/api-rest/src/RestClient.ts:196]: @elorzafe It's from the original implementation. My understanding it this prevents repeated call to `this.Credentials.get()` and reduce the duplicated debug messages caused by starting from beginning(`this.ajax`)

**Review by manueliglesias — APPROVED** [review]
LGTM

**Review by elorzafe — APPROVED** [review]
Looks good, thanks @AllanZhengYP 🏅

**Review by david-mcafee — APPROVED** [review]
LGTM! 🎉

**Pull request merged**